src/util/hooks.js

```diff
@@ -0,0 +1,60 @@
+// @flow
+
+import * as React from 'react'
+
+type SetState<S> = (value: S | ((state: S) => S)) => void
+
+type UseCallback = <T>(callback: T, deps: any[]) => T
+
+type UseContext = <T>(context: React.Context<T>) => T
+
+type UseDebugValue = <T>(value: T, format?: (value: T) => any) => void
+
+type UseEffect = (effect: () => void | (() => void), deps?: any[]) => void
+
+type UseImperativeHandle = (ref: any, init: () => any, deps?: any[]) => void
+
+type UseMemo = <T>(init: () => T, deps?: any[]) => T
+
+type UseReducer = {
+  // Normal version:
+  <State, Action>(reducer: (state: State | void, action: Action) => State, init: State | void): [State, (action: Action) => void],
+
+  // Initializer version:
+  <State, Action, Init>(
+    reducer: (state: State | void, action: Action) => State,
+    init: Init,
+    initializer: (init: Init) => State
+  ): [State, (action: Action) => void]
+}
+
+type UseRef = {
+  // Value container:
+  <T>(init: T): { current: T },
```

Comment:
Can you merge this two into one type or having the readonly only on off them really matters?
If not, can't you do something like this? 
`type UseRef = <T>(init?: T): { +current: T | null }`

Reply:
This would be incorrect, because in the first case, `current` will not be `null`, and requiring a bunch of `null` tests would be annoying. In the second case, it totally can be `null`, because React links it to the mounted / unmounted lifetime of a component. The other difference is that in case 1, the user is allowed to write to `current`, but in the second case `current` is read-only (because React manages it).

Reply:
After testing this extensively & hitting my head against a wall, I do believe that the overload is still needed, but the React side did need some changes. With these fixes, you can either do:

```
const ref: { current: null | typeof SomeComponent } = useRef()
return <SomeComponent ref={ref} />
```

Or you can do:

```
const ref: { current: string } = useRef('')
```

Getting both cases to work at the same time is tricky, and I think the overload may be the only way to do it. The overload is *certainly* necessary in Typescript, so it stands to reason Flow would need it too.

Reply:
👍🏼